torchinductor/scheduler.py

```diff
@@ -459,87 +537,237 @@ def create_fx_from_buffers(nodes, fname, print_graph=False):
+        if isinstance(snode, FusedSchedulerNode):
+            for x in snode.snodes:
+                buf_to_fx_node[x.get_name()] = fx_node
+        buf_to_fx_node[name] = fx_node
 
-        name_to_fx_node[name] = fx_node
         if first_node is None:
             first_node = fx_node
 
     # create edges between nodes
-    for snode in nodes:
+    for snode in snodes:
         node = snode.node
-        name = node.get_name()
-        deps = node.get_reads()
-        fx_node = name_to_fx_node[name]
+        name = snode.get_name()
+        if isinstance(snode, FusedSchedulerNode):
+            deps = [x.node.get_reads() for x in snode.snodes]
+            deps = [x for x in itertools.chain.from_iterable(deps)]
+        else:
+            deps = node.get_reads()
+        fx_node = buf_to_fx_node[name]
 
         new_args = []
         for dep in deps:
-            if dep.name in name_to_fx_node:
-                dep_node = name_to_fx_node[dep.name]
+            if dep.name in buf_to_fx_node:
+                dep_node = buf_to_fx_node[dep.name]
             else:
                 with graph.inserting_before(first_node):
                     dep_node = graph.placeholder(dep.name)
-                    name_to_fx_node[dep.name] = dep_node
+                    buf_to_fx_node[dep.name] = dep_node
             new_args.append(dep_node)
 
         fx_node.args = tuple(new_args)
 
-    outputs = []
-    for _, v in name_to_fx_node.items():
-        if len(v.users) == 0:
-            outputs.append(v)
     graph.output(outputs[0] if len(outputs) == 1 else tuple(outputs))
+    return graph
 
-    if print_graph:
-        print(graph)
-    print("starting creating module")
-    gm = GraphModule({}, graph)
-    graph = legalize_graph(gm)
-    gm.graph.lint()
-    print("starting drawing")
-    draw_graph(gm, fname, clear_meta=False)
+
+def prefuse_nodes(snodes: List[BaseSchedulerNode]):
+    """
+    Groups together fusible scheduler nodes into a FusedSchedulerNode
+    Note: O(N^2) asymptotics
+    """
+    if len(snodes) == 0:
+        return snodes
+    from torch.fx.passes.tools_common import legalize_graph
+
+    def toposort(graph):
+        gm = fx.GraphModule({}, graph)
+        legalize_graph(gm)
+        return gm.graph
+
+    graph = create_fx_from_snodes(snodes)
+    graph = toposort(graph)
+
+    for idx, node in enumerate(graph.nodes):
+        node.order = idx
+
+    fusible_nodes = [
+        node
+        for node in graph.nodes
+        if node.op == "call_function" and node.meta["fusion_meta"].type == "compute"
+    ]
+
+    def is_fusible(src, dst):
+        # Finds whether there's a path from src to dst that isn't a direct edge
+        cur_nodes = collections.deque(
+            [user for user in src.users if user != dst]
+            + [user for user in dst.users if user != src]
+        )
+        vis = set()
+        while len(cur_nodes) > 0:
+            cur = cur_nodes.popleft()
+            if cur in vis:
```

Comment:
Small perf optimization that uses topological sorting.
```suggestion
            if cur in vis or cur.order > max(src.order, dst.order):
```

Reply:
I actually had this perf optimization before, but removed it due to some complexity around correctness.

I’m not actually sure your suggestion maintains correctness. Lemme comment more in that part.